app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,347 @@
+                        struct zmk_behavior_binding_event event) {
+    // Start the timer if one hasn't started already. Reset it if one exists.
+    tap_dance->release_at = event.timestamp + tap_dance->config->tapping_term_ms;
+    int32_t ms_left = tap_dance->release_at - k_uptime_get();
+    if (ms_left > 0) {
+        k_delayed_work_submit(&tap_dance->release_timer, K_MSEC(ms_left));
+    }
+}
+
+// press_tap_dance_behavior
+
+// Returns behavior_keymap_binding_pressed to ensure that keycode is actually pressed
+static inline int press_tap_dance_behavior(struct active_tap_dance *tap_dance, int64_t timestamp) {
+    // Press chosen tap dance binding
+    struct zmk_behavior_binding binding =
+        (tap_dance->counter < tap_dance->config->behavior_count)
+            ? tap_dance->config->behaviors[(tap_dance->counter) - 1]
+            : tap_dance->config->behaviors[(tap_dance->config->behavior_count) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    return behavior_keymap_binding_pressed(&binding, event);
+}
+
+// release_tap_dance_behavior
+
+// Returns behavior_keymap_binding_pressed to ensure that keycode is actually released
+static inline int release_tap_dance_behavior(struct active_tap_dance *tap_dance,
+                                             int64_t timestamp) {
+    // Release chosen tap dance binding
+    struct zmk_behavior_binding binding =
+        (tap_dance->counter < tap_dance->config->behavior_count)
+            ? tap_dance->config->behaviors[(tap_dance->counter) - 1]
+            : tap_dance->config->behaviors[(tap_dance->config->behavior_count) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    clear_tap_dance(tap_dance);
+    return behavior_keymap_binding_released(&binding, event);
+}
+
+// on_tap_dance_binding_pressed
+
+// Code that determines what happens when a tap-dance binding is pressed.
+// Does not register a keycode press unless the tap-dance's counter is
+// equal to the maximum number of bindings associated with it.
+// E.g. if a defined tap-dance has bindings = <&kp N1>, <&kp N2>, <&kp N3>;
+// when the counter is equal to 3, press_tap_dance_behavior will be called
+// to register N3's keycode being pressed
+static int on_tap_dance_binding_pressed(struct zmk_behavior_binding *binding,
+                                        struct zmk_behavior_binding_event event) {
+    // Code executes when the tap dance key is pressed.
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_tap_dance_config *cfg = dev->config;
+    struct active_tap_dance *tap_dance;
+    tap_dance = find_tap_dance(event.position);
+    // No active tap dance found on keypress location, create a new one
+    if (tap_dance == NULL) {
+        tap_dance = new_tap_dance(event.position, cfg);
```

Comment:
Needs some error checking in case we hit the max held.

Reply:
`static int new_tap_dance`, formerly `static struct active_tap_dance *new_tap_dance` already had error checking for the max held. I've added a few extra comments and another `LOG_ERR` statement to make this clear. 
```
static int new_tap_dance(uint32_t position,
                         const struct behavior_tap_dance_config *config,
                         struct active_tap_dance **tap_dance) {
    // Attempt to create a new instance of a tap-dance
    for (int i = 0; i < ZMK_BHV_TAP_DANCE_MAX_HELD; i++) {
        struct active_tap_dance *const tap_dance = &active_tap_dances[i];
        // Skip any elements in active_tap_dances[] that should not be altered (active/paused tap-dances)
        if (tap_dance->position != ZMK_BHV_TAP_DANCE_POSITION_FREE || tap_dance->timer_cancelled) {
            continue;
        }
        tap_dance->counter = 1;
        tap_dance->position = position;
        tap_dance->config = config;
        tap_dance->release_at = 0;
        tap_dance->is_pressed = true;
        tap_dance->timer_started = true;
        tap_dance->timer_cancelled = false;
        tap_dance->tap_dance_decided = false;
        return 0;
    }
    // Unable to store new tap-dance in array active_tap_dances[].
    LOG_ERR("Unable to store new tap-dance. Insufficient space.");
    return -ENOMEM;
}
```

Reply:
I realize that the `for` loop could just as easily be rewritten with flipped logic. Something like:
```
static int new_tap_dance(uint32_t position,
                         const struct behavior_tap_dance_config *config,
                         struct active_tap_dance **tap_dance) {
    for (int i = 0; i < ZMK_BHV_TAP_DANCE_MAX_HELD; i++) {
        struct active_tap_dance *const tap_dance = &active_tap_dances[i];
        if (tap_dance->position == ZMK_BHV_TAP_DANCE_POSITION_FREE) {
            // Found a free space in active_tap_dances[], create new tap-dance
            tap_dance->counter = 1;
            tap_dance->position = position;
            tap_dance->config = config;
            tap_dance->release_at = 0;
            tap_dance->is_pressed = true;
            tap_dance->timer_started = true;
            tap_dance->timer_cancelled = false;
            tap_dance->tap_dance_decided = false;
            return 0;
        }
    }
    // Unable to store new tap-dance in array active_tap_dances[].
    LOG_ERR("Unable to store new tap-dance. Insufficient space.");
    return -ENOMEM;
}
```
The `if` statement in the loop shouldn't need to consider `tap_dance->timer_cancelled` because `clear_tap_dance` doesn't alter the value of `tap_dance->timer_cancelled`. It's implied that the status of the timer is irrelevant if the position of the tap-dance is `ZMK_BHV_TAP_DANCE_POSITION_FREE`.

Reply:
Maybe I wasn't clear, I mean that this code calls `new_tap_dance` but doesn't actually verify it was given back a new tap dance struct to work with.

Reply:
I understand. I'll keep that in mind for today's revisions.